core/src/main/kotlin/org/evomaster/core/problem/graphql/GraphQLUtils.kt

```diff
@@ -165,4 +170,273 @@ object GraphQLUtils {
+                                            history.remove(ob)
+                                        } else {
+                                            history.remove(ob)
+                                        }
+                                    }
+                                    history.remove(element.tableFieldType)// remove the interface type object to the history:the normal case (not cycle)
+                                } else {
+                                    history.remove(element.tableFieldType)// remove the interface type object to the history:the cycle case
+                                }
+                            }
+                        }
+                    }
+
+                }
+            }
+        }
+        return graph
+    }
+
+    private fun addNode(node: String, graph: MutableMap<String, GraphInfo>): GraphInfo? {
+        return graph.putIfAbsent(node, GraphInfo())
+    }
+
+    private fun addEdge(source: String, destination: String, fieldName: String, graph: MutableMap<String, GraphInfo>): Boolean? {
+        addNode(source, graph)
+        addNode(destination, graph)
+        graph[source]?.edges?.add(destination)
+        return graph[source]?.fields?.add(fieldName)
+    }
+
+    /**
+     * The number of nodes
+     * Note: the count include Query and Mutation
+     */
+    fun getGraphSize(graph: MutableMap<String, GraphInfo>): Int {
+        return graph.size
+    }
+
+    /**
+     * The number of fields that are pointers to other nodes in the graph.
+     * Note: fields in Query and Mutation are considered in this count.
+     */
+    fun getNbrOfEdges(graph: MutableMap<String, GraphInfo>): Int {
+        return graph.values.stream().mapToInt { it.edges.size }.sum()
+    }
+
+    /**
+     * The number of fields in the type Query or mutation
+     */
+    fun getNbrQueriesOrMutations(queriesMutationsEntryPoint: String, graph: MutableMap<String, GraphInfo>): Int? {
+        return graph[queriesMutationsEntryPoint.toLowerCase()]?.fields?.size
+    }
+
+    /**
+     * Get the number of fields: the number of fields in all nodes, excluding Query and Mutation
+     */
+    fun getNbrFields(graph: MutableMap<String, GraphInfo>): MutableList<Int> {
+        val nbrFields: MutableList<Int> = mutableListOf()
+        for (element in graph.entries) {
+            if (element.key == "query" || element.key == "querytype" || element.key == "root" || element.key == "mutation")
+                continue
+            else
+                nbrFields.add(element.value.fields.size)
+
+        }
+        return nbrFields
+    }
+
+    /**
+     * Get all adjacent vertex
+     */
+    fun getAdjacent(vertex: String, graph: MutableMap<String, GraphInfo>): MutableSet<String>? {
+        return graph[vertex]?.edges
+    }
+
+    /**
+     * Check the existence of a node in the graph
+     */
+    fun checkNode(node: String, graph: MutableMap<String, GraphInfo>): Boolean {
```

Comment:
`checkNodeExists`.
and like in many other function, read-only methods should take as input a read-only version of the graph, ie, a `Map` instead of a `MutableMap` (this latter extends `Map` in Kotlin)

Reply:
-Done